solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java

```diff
@@ -0,0 +1,356 @@
+        activeShardRequests.computeIfAbsent(
+            shardRequest,
+            k -> {
+              List<ShardRequestActor> actors = new ArrayList<>();
+              actors.add(new NodeStatsCollector(slowNodeDetector));
+              if (shardsTolerant) {
+                actors.add(
+                    new SlowNodeTimeoutActor(
+                        slowNodeTimeout, dryRun, slowNodeDetector.getSlowNodes(), timeoutCallback));
+              }
+              return new ShardRequestTracker(actors);
+            });
+
+    tracker.actors.forEach(actor -> actor.onRequestSubmitted(shardUrls, future));
+    tracker.outstandingRequestCount.incrementAndGet();
+
+    return new ShardRequestTrackingCallback(
+        future, shardRequest, tracker, shardsTolerant, shardUrls);
+  }
+
+  /**
+   * Callback to track ShardRequest and a Future created by sending such request to a single shard
+   * URL.
+   *
+   * <p>Take note that the ShardRequest can be reused/shared by sending to many different shards.
+   * Therefore, we need this class as a reference back to the corresponding ShardRequest when such
+   * future is completed.
+   */
+  class ShardRequestTrackingCallback implements HttpShardHandler.ShardRequestCallback {
+    private final ShardRequest shardRequest;
+    private final Future<LBSolrClient.Rsp> future;
+    private final ShardRequestTracker tracker;
+    private final List<String> shardUrls;
+    private final boolean shardsTolerant;
+
+    ShardRequestTrackingCallback(
+        Future<LBSolrClient.Rsp> future,
+        ShardRequest shardRequest,
+        ShardRequestTracker tracker,
+        boolean shardsTolerant,
+        List<String> shardUrls) {
+      this.future = future;
+      this.shardRequest = shardRequest;
+      this.tracker = tracker;
+      this.shardUrls = shardUrls;
+      this.shardsTolerant = shardsTolerant;
+    }
+
+    @Override
+    public void onResponse(LBSolrClient.Rsp response, long elapsedTime) {
+      onComplete(elapsedTime, response.getServer(), false);
+    }
+
+    @Override
+    public void onException(Throwable exception, long elapsedTime) {
+      // TODO is it possible to infer the selected node? If it has timed out, perhaps we can assume
+      // all shardNodes are slow?
+      onComplete(elapsedTime, null, true);
+    }
+
+    private void onComplete(long elapsedTime, String selectedShardUrl, boolean isException) {
+      try {
+        tracker.actors.forEach(
+            actor -> actor.onRequestCompleted(selectedShardUrl, shardUrls, future, elapsedTime));
+        int outstandingRequestCount = tracker.outstandingRequestCount.decrementAndGet();
+        if (isLastResponse(isException, outstandingRequestCount)) {
+          tracker.actors.forEach(ShardRequestActor::close);
+
+          // there could be some race condition here but benign
+          // (other thread triggers onRequestSubmit on actors which are already closed etc)
+          activeShardRequests.remove(shardRequest);
```

Comment:
I think the race condition on `activeShardRequests.remove(shardRequest)` is benign; but there's also a possibility here that `tracker.actors.forEach(ShardRequestActor::close)` will be called multiple times for the same tracker/actors.

Reply:
As discussed in Slack, we will:
1. Keep the current flow - `computeIfAbsent` in `onRequestSubmit`, no extra synchronization in `onComplete`
2. Relax the requirement in `ShardRequestActor` javadoc, that the implementation should prepare that more method invocations could arrive after `close` is invoked (though unlikely), and they should take care to NOT throw exception/cause issues, the actual behavior is up to the implementation